protocol.go

```diff
@@ -0,0 +1,451 @@
+	}
+
+	second := p.wq[:count-length]
+
+	emptyBufferIndices(first)
+	emptyBufferIndices(second)
+}
+
+func (p *Protocol) Read(header PacketHeader, buf []byte, transmit transmitFunc) error {
+	p.readAckBits(header.ACK, header.ACKBits)
+
+	if !header.Unordered && !p.trackRead(header.Sequence) {
+		return nil
+	}
+
+	p.trackUnacked()
+
+	if err := p.writeAcksIfNecessary(transmit); err != nil {
+		return fmt.Errorf("failed to write acks when necessary: %w", err)
+	}
+
+	if header.Empty {
+		return nil
+	}
+
+	if p.ph != nil {
+		p.ph(p.addr, header.Sequence, buf)
+	}
+
+	//log.Printf("%s: recv    (seq=%05d) (ack=%05d) (ack_bits=%032b) (size=%d) (reliable=%t)", p.conn.LocalAddr(), header.Sequence, header.ACK, header.ACKBits, len(buf), !header.Unordered)
+
+	return nil
+}
+
+func (p *Protocol) createAckIfNecessary() (header PacketHeader, needed bool) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	lui := p.lui
+
+	for i := uint16(0); i < ACKBitsetSize; i++ {
+		if p.rq[(lui+i)%uint16(len(p.rq))] != uint32(lui+i) {
+			return header, needed
+		}
+	}
+
+	lui += ACKBitsetSize
+	p.lui = lui
+	p.ls = time.Now()
+
+	p.waitUntilReaderAvailable()
+
+	header.Sequence, header.ACK = p.nextWriteIndex(), lui-1
+	header.ACKBits = p.prepareAckBits(header.ACK)
+	header.Empty = true
+
+	needed = !p.die
+
+	return header, needed
+}
+
+func (p *Protocol) writeAcksIfNecessary(transmit transmitFunc) error {
+	for {
+		header, needed := p.createAckIfNecessary()
+		if !needed {
+			return nil
+		}
+
+		//log.Printf("%s: ack     (seq=%05d) (ack=%05d) (ack_bits=%032b)", p.conn.LocalAddr(), header.Sequence, header.ACK, header.ACKBits)
+
+		if err := p.write(header, nil, transmit); err != nil {
+			return fmt.Errorf("failed to write ack packet: %w", err)
+		}
+	}
+}
+
+func (p *Protocol) readAckBits(ack uint16, ackBits uint32) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	for idx := uint16(0); idx < ACKBitsetSize; idx, ackBits = idx+1, ackBits>>1 {
+		if ackBits&1 == 0 {
+			continue
+		}
+
+		i := (ack - idx) % uint16(len(p.wq))
+		if p.wq[i] != uint32(ack-idx) || p.wqe[i].acked {
+			continue
+		}
+
+		if p.wqe[i].buf != nil {
+			p.pool.Put(p.wqe[i].buf)
+		}
+
+		p.wqe[i].buf = nil
+		p.wqe[i].acked = true
+	}
+}
+
+func (p *Protocol) trackRead(idx uint16) bool {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	i := idx % uint16(len(p.rq))
+
+	if p.rq[i] == uint32(idx) { // duplicate packet
+		return false
+	}
+
+	if seq.GT(idx+1, p.ri) {
+		p.clearReads(p.ri, idx)
+		p.ri = idx + 1
+	}
+
+	p.rq[i] = uint32(idx)
+
+	return true
+}
+
+func (p *Protocol) clearReads(start, end uint16) {
+	count, size := end-start+1, uint16(len(p.rq))
+
+	if count >= size {
+		emptyBufferIndices(p.rq)
+		return
+	}
+
+	first := p.rq[start%size:]
+	length := uint16(len(first))
+
+	if count <= length {
+		emptyBufferIndices(first[:count])
+		return
+	}
+
+	second := p.rq[:count-length]
+
+	emptyBufferIndices(first)
+	emptyBufferIndices(second)
+}
+
+func (p *Protocol) trackAcked(ack uint16) {
+	p.mu.Lock()
```

Comment:
Would you be able to double-check where mutex locks might not be necessary and remove them? I have a feeling that mutices are locked/unlocked too often.

Reply:
I did following changes.
Please make checkbox marked if collect.

- [ ] I moved mutex of `retransmitUnackedPackets` for better performance
- [ ] I deleted mutex of `(p *Protocol) close()` because `p.die` is assigned only here
- [ ] I deleted mutex of `trackUnacked` because `p.oui` is assigned only here and `trackUnacked` is only called from `(p *Protocol) Read(`
- [ ] I deleted mutex of `trackRead` because `p.rq` is assigned only here and `trackRead` is only called from `(p *Protocol) Read(`

Reply:
Hm, would you be able to make a few unit tests to prove that no race conditions will happen with your changes?

Reply:
I added test code. 
https://github.com/tak1827/reliable/blob/feature/conn-struct-decoupling/endpoint_test.go#L182

I'm no idea how to test race conditions💦
So, I simply simulate write and read heavy condition by sending packet concurrently.

Reply:
4 writes in parallel is working.
The error came from calling side race condition. So I modified calling side.